### ISSUE TITLE
in_http is replacing '+' symbol in JSON fields

### ISSUE THREAD

**Issue by elelement** [fluent/fluentd#1291]
td-agent version: td-agent 0.12.26
Environment information, e.g. OS: Debian 8 for td-agent, Windows 7 for Python code.
Plugin: in_http.rb
Problem explanation:

I have observed a possible bug in in_http plugin. I have a HTTP source redirecting logs to a Graylog server and to stdout. Graylog was rejecting my messages because the datetime format was incorrect. I realized that the '+' symbol of the time zone was being replaced by a space. I think it was only a date format issue, but I've also tried with other fields, like user name and the problem persist.

In the following image you can see in the top of the image, the td-agent.log trace after being redirected to stdout and in the bottom, a tcpdump trace listening on port 8888. 

![error_fluent_http](https://cloud.githubusercontent.com/assets/449705/19652351/c16ed210-9a10-11e6-8369-49c035f2f9d5.PNG)

As you can see, the data is correctly sent to the in_http plugin, but both graylog and stdout outputs are wrong.

In summary: the observed behavior is that in_http replaces '+'  by ' '.

Thanks for your support.

**Comment by repeatedly** [comment]
How about using "multipart/form-data" instead of `application/x-www-form-urlencoded`?

**Comment by elelement** [comment]
If I use `multipart/form-data`, I get a 

> 400 Bad Request
> 'json' or 'msgpack' parameter is required

Please, take a look at tcpdump trace:
```
POST /graylog.http.dev HTTP/1.1
Host: <my_ip>:8888
Connection: keep-alive
Accept-Encoding: gzip, deflate
Accept: */*
User-Agent: python-requests/2.11.1
Content-type: multipart/form-data
Content-Length: 186


write error to stdout

json={"host":"es-mad-wkdw1417","level":"INFO","username":"jmcc","path":"(40) - test.py","datetime":"2016-10-25T10:02:50+0000","full_message":"","message":"HTTP info trace"}
write error to stdout

HTTP/1.1 400 Bad Request
Content-type: text/plain
Connection: Keep-Alive
Content-length: 58

400 Bad Request
'json' or 'msgpack' parameter is required
```

Another example using a PHP script:

```
POST /graylog.http.dev HTTP/1.1
Host: <my_ip>:8888
Accept: */*
Content-Type: multipart/form-data
Content-Length: 274

json={"host":"dashboardjz","port":8888,"datetime":"2016-10-25T10:36:13+02:00","username":"adiaz+","path":"\/var\/www\/html\/test\/adiaz\/src\/testSyslog.php","message":"Mensaje corto level EMERGENCY","full_message":"Mensaje completo con todo el detalle","level":"EMERGENCY"}
write error to stdout

HTTP/1.1 400 Bad Request
Content-type: text/plain
Connection: Keep-Alive
Content-length: 58

400 Bad Request
'json' or 'msgpack' parameter is required
```

Same result using `text/plain`, but `application/x-www-form-urlencoded` works fine except for the '+' replacement.

**Comment by repeatedly** [comment]
I tested your `datetime` value but it seems no problem.

```
% fluentd -c http_debug.conf
2016-10-26 05:29:19 +0900 [info]: reading config file path="http_debug.conf"
2016-10-26 05:29:19 +0900 [info]: starting fluentd-0.12.29
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-mixin-config-placeholders' version '0.3.1'
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-plugin-beats' version '0.1.1'
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-plugin-elasticsearch' version '1.7.0'
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-plugin-flowcounter-simple' version '0.0.4'
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-plugin-record-modifier' version '0.5.0'
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-plugin-s3' version '0.7.1'
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-plugin-secure-forward' version '0.4.2'
2016-10-26 05:29:19 +0900 [info]: gem 'fluent-plugin-td' version '0.10.29'
2016-10-26 05:29:19 +0900 [info]: gem 'fluentd' version '0.12.29'
2016-10-26 05:29:19 +0900 [info]: adding match pattern="debug.*" type="stdout"
2016-10-26 05:29:19 +0900 [info]: adding source type="http"
2016-10-26 05:29:19 +0900 [info]: using configuration file: <ROOT>
  <source>
    @type http
    port 8888
  </source>
  <match debug.*>
    @type stdout
  </match>
</ROOT>
2016-10-26 05:29:21 +0900 debug.test: {"k":"v","datetime":"2016-10-25T10:02:50+0000"}
```

- client code : Ruby

```
require 'net/http'
require 'json'

url="http://127.0.0.1:8888/debug.test"
json = {"k" => "v","datetime" => "2016-10-25T10:02:50+0000"}.to_json
result = Net::HTTP.post_form(URI.parse(url), {'json' => json})
```

I assume you use don't encode json parameter correctly.

**Comment by repeatedly** [comment]
In GET case, 

```
URI.encode_www_form({json: {"k" => "v","datetime" => "2016-10-25T10:02:50+0000"}.to_json})
```

This code is OK,

```
"json=#{{"k" => "v","datetime" => "2016-10-25T10:02:50+0000"}.to_json}"
```

but this is NG

**Comment by elelement** [comment]
:(  I should have realized it, sorry. Just for the record, it didn't worked for me directly encoding the whole record, but just the string in the ```json{%s}``` form. 

This works:

```python
return requests.post('''http://%s:%d/%s''' % (self.address[0], self.address[1], self.resource), 
    '''json={%s}''' % urllib.quote(log_entry), 
    headers={"Content-type": "application/x-www-form-urlencoded; charset=UTF-8"}).content
```

But this doesn't:
```python
return requests.post('''http://%s:%d/%s''' % (self.address[0], self.address[1], self.resource), 
    urllib.quote('''json={%s}''' % log_entry), 
    headers={"Content-type": "application/x-www-form-urlencoded; charset=UTF-8"}).content
```
Maybe it is useless or already known to mention the content type in the docs, but it would be nice in case someone asks itself how the request should look like.

Thank you so much for your time and apologize for the issue.
